### ISSUE TITLE
Inconsistency in ak.concatenate

### ISSUE THREAD

**Issue by grst** [scikit-hep/awkward#1586]
### Version of Awkward Array

1.9.0rc10

### Description and code to reproduce

When concatenating two regular arrays, I would expect the output array to be regular, too. 

This works in this case
```python
import awkward._v2 as ak
import numpy as np

np1 = ak.Array(np.array([[53,274],[39,388]]))
np1
# <Array [[53, 274], [39, 388]] type='2 * 2 * int64'>
np2 = ak.Array(np.array([[53,274],[39,388]]))
np2
# <Array [[53, 274], [39, 388]] type='2 * 2 * int64'>

ak.concatenate([np1, np2])
# <Array [[53, 274], [39, 388], [53, ...], [39, 388]] type='4 * 2 * int64'>
```

However, when using the same arrays, but generated from json, the second dimension is variable after concatenation

```python
a1 = ak.from_json('[[53,274],[39,388]]')
a2 = ak.from_json('[[53,274],[39,388]]')
a1 = ak.to_regular(a1, 1)
a2 = ak.to_regular(a2, 1)
a1
# <Array [[53, 274], [39, 388]] type='2 * 2 * int64'>
a2 
# <Array [[53, 274], [39, 388]] type='2 * 2 * int64'>

ak.concatenate([a1, a2])
# <Array [[53, 274], [39, 388], [53, ...], [39, 388]] type='4 * var * int64'>
```

### Context: 
Discovered while working on https://github.com/scverse/anndata/pull/647#issuecomment-1061847089. The unittest for anndata concatenation doesn't work because of this. I can work around this by manually making the array regular again after concatenation, but wanted to check if this is something that's going to be fixed on the awkward array side first.

**Comment by agoose77** [comment]
So, the technical cause for this is that 2D NumPy arrays (the `NumpyArray` layout node) are treated differently to 2D views over 1D NumPy Arrays (the `RegularArray` layout node). Somewhere in the concatenation logic we're perhaps losing the regular dimension to broadcasting, or something similar. This looks like a bug with a simple fix to me.

**Comment by grst** [comment]
I confirm the example described above is fixed! However... I found that when concatenating along another axis, the resulting array is still not regular: 

```python
a1 = ak.from_json('[[53,274],[39,388]]')
a2 = ak.from_json('[[53,274],[39,388]]')
a1 = ak.to_regular(a1, 1)
a2 = ak.to_regular(a2, 1)

# this works now
ak.concatenate([a1, a2])
# <Array [[53, 274], [39, 388], [53, ...], [39, 388]] type='4 * 2 * int64'>

# this doesn't
ak.concatenate([a1, a2], axis=1)
# <Array [[53, 274, 53, 274], [39, 388, ..., 388]] type='2 * var * int64'>
```

Sorry for all those nasty edge-cases!

**Comment by jpivarski** [comment]
FYI, these aren't bug-fixes so much as new implementations. The inconsistency came from the fact that sometimes regular dimensions are represented by RegularArray nodes and sometimes (if they're at leaf-level) by NumpyArray nodes with `ndim > 1`. When the concatenation is applied to the NumpyArray nodes, we let NumPy take care of it, and naturally regular inputs beget regular outputs, but when concatenation is applied to RegularArray nodes, they were first converted to variable-length ListOffsetArray nodes as a matter of policy. Since the distinction between RegularArray and NumpyArray is not user-visible (unless you delve into `.layout`), this is an inconsistent policy. That's what I'm fixing: I'm giving the RegularArrays a specialized path.

On the other hand, that means that this is also an optimized path. When concatenating RegularArrays, these fixes prevent us from wasting time creating ListOffsetArrays that just happen to be regular (`offsets` that can be generated by `np.arange`). While I had all of this open, I applied a new constraint to `ak.concatenate` with `axis=0`, that if the Forms[^1] of all inputs are identical, the output would have that Form as well (#1605), which is something that will help Dask because it often has to concatenate computed partitions, and would like to treat the concatenated partitions in the same way as a single partition.

[^1]: a Form is a detailed type: `array.type` is high-level (abstract types), `array.layout.form` is low-level (concrete types, enough detail for Numba-compilation, for instance).

**Comment by jpivarski** [comment]
#1609 fixes the `axis != 0` case as well. Concatenation naturally splits into an `axis = 0` case (fixed by #1604) and an `axis != 0` case (fixed by #1609), so we should be done now.

I also added tests for cases in which one side of a concatenation is a RegularArray and another is a NumpyArray, for both `axis = 0` and `axis != 0`.

One thing, though: if you have an option-regular type (`option[100 * xyz]`), that still concatenates to variable-length outputs because any missing lists are treated like empty lists, which have a different length than the regular size (`100` in my example).

**Comment by grst** [comment]
That was quick! Thank you so much, also for the detailed explanations!
